### ISSUE TITLE
InsertLeave: luasnip/lua/luasnip/util/mark.lua:36: index out of range

### ISSUE THREAD

**Issue by oxalica** [L3MON4D3/LuaSnip#239]
I'm trying nvim's builtin LSP with `nvim-cmp` and `luasnip`. I follow [`nvim-cmp`'s README](https://github.com/hrsh7th/nvim-cmp/#recommended-configuration`) and use `luasnip` as snippet engine.

When I do some editing, run a tool outside nvim changing related files (like `git checkout`) and go back to nvim, the file is autoreloaded, but this error begins to occur every time I leave the insert mode.
```
Error detected while processing InsertLeave Autocommands for "*":
E5108: Error executing lua ...ack/home-manager/start/luasnip/lua/luasnip/util/mark.lua:36: index out of range
stack traceback:
        [C]: in function 'str_utfindex'
        ...ack/home-manager/start/luasnip/lua/luasnip/util/mark.lua:36: in function 'pos_end'
        ...home-manager/start/luasnip/lua/luasnip/nodes/snippet.lua:440: in function 'enter_node'
        ...dir/pack/home-manager/start/luasnip/lua/luasnip/init.lua:322: in function 'active_update_dependents'
        [string ":lua"]:1: in main chunk
```
I'm not sure if it's an issue of `luasnip` or `nvim-cmp`, but it looks like some cache not being invalidated during autoreload.

Versions:
`luasnip`: 577045e9adf325e58f690f4d4b4a293f3dcec1b3
`nvim-cmp`: af07ff9b7973e95eff9e0275e13fe0350281208b
`cmp_luasnip`: 75bf6434f175206cd219f9d2bbcae154a009346c

**Comment by L3MON4D3** [comment]
Yep, that's luasnip. When the files are changed outside of vim the extmarks we use to mark node-positions stay at their positions (seems like that at least), but the positions could very well be outside the new text.

could you try if the `out_of_range_fixes` branch helps with this issue?

**Comment by L3MON4D3** [comment]
I merged the changes, please reopen if you still get errors

**Comment by oxalica** [comment]
Thanks for the quick fix. I've been testing it for some hours and have not seen any errors yet.

**Comment by L3MON4D3** [comment]
Nice, that's great to hear :D
